Fix #1560: add linking related UX in ToC pane

roland-ruedenauer opened the pull request:
The context menu for ToC entries gets extended by offering now in addition
- "Copy link to heading"
- "Open in New Window"

The PR also fixes the wrong displaying of headings with an explicit anchor in the ToC pane (filtering for a tag was added bc of this).

introt commented:
Nice work!

* On the first run I had an issue with the menu items being off-by-one which I couldn't reproduce.
* If you have the time, it'd imo make sense to disable pro-/demoting of horizontal bars
- [ ] Pasting a link to a page title heading broke with the error shown below. This works on latest `develop` commit when copied from inside the wiki page, but is broken here regardless of whether copying from the page or the plugin.

```
DEBUG: set_pagelink <Notebook: Test> <Page: test:subtest:subsub> anchor=subsub text=subsub
DEBUG: Action: paste
DEBUG: Targets available for paste: [Gdk.Atom.intern("TIMESTAMP", False), Gdk.Atom.intern("TARGETS", False), Gdk.Atom.intern("MULTIPLE", False), Gdk.Atom.intern("SAVE_TARGETS", False), Gdk.Atom.intern("UTF8_STRING", False), Gdk.Atom.intern("COMPOUND_TEXT", False), Gdk.Atom.intern("TEXT", False), Gdk.Atom.intern("STRING", False), Gdk.Atom.intern("text/plain;charset=utf-8", False), Gdk.Atom.intern("text/plain", False)], we want parsetree
DEBUG: Requesting data for ('text/x-zim-parsetree', 0, 1) -- using set_with_data() workaround
Traceback (most recent call last):
  File "/home/introt/repos/zim-desktop-wiki/zim/gui/pageview/__init__.py", line 4141, in do_paste_clipboard
    self.get_buffer().paste_clipboard(Clipboard, None, self.get_editable(), text_format=format)
  File "/home/introt/repos/zim-desktop-wiki/zim/gui/pageview/__init__.py", line 3376, in paste_clipboard
    parsetree = clipboard.get_parsetree(self.notebook, self.page, text_format)
  File "/home/introt/repos/zim-desktop-wiki/zim/gui/clipboard.py", line 641, in get_parsetree
    return parsetree_from_selectiondata(selectiondata, notebook, path, text_format)
  File "/home/introt/repos/zim-desktop-wiki/zim/gui/clipboard.py", line 228, in parsetree_from_selectiondata
    replace_parsetree_links_and_copy_images(parsetree, notebook, path)
  File "/home/introt/repos/zim-desktop-wiki/zim/notebook/content_updater.py", line 94, in replace_parsetree_links_and_copy_images
    parsetree.replace(
  File "/home/introt/repos/zim-desktop-wiki/zim/formats/__init__.py", line 668, in replace
    self._replace(self._etree.getroot(), tags, func)
  File "/home/introt/repos/zim-desktop-wiki/zim/formats/__init__.py", line 679, in _replace
    replacement = func(Element.new_from_etree(child))
  File "/home/introt/repos/zim-desktop-wiki/zim/notebook/content_updater.py", line 168, in _replace_links_to_page_and_copy_images
    return notebook._update_link_tag(node, new_path, target, oldhref)
  File "/home/introt/repos/zim-desktop-wiki/zim/notebook/notebook.py", line 842, in _update_link_tag
    short = newhref.parts()[-1]
IndexError: list index out of range
```

roland-ruedenauer commented:
@introt thanks for testing the changes so quickly. You surprised me.
Unfortunately I wasn't able to reproduce the problem you ran into in my local testing. I tested with having horizontal bars displayed in the ToC pane and without. Can you maybe have a look what's contained in new_href when the exception is raised or provide your test page content?
I also wonder how I could have broken copy & paste of links to headings from the page view directly.
Did you copy from the ToC pane immediately after changing the page content? In this case I could imagine we might run into problems, because there's a significant delay before the ToC get's updated after you modify the page content.

roland-ruedenauer commented:
While testing my changes with the new release 74.1 I observed a few minor issues
* ToC should always be above other TextView sibling windows. Having a source-view item on the page the ToC pane on first display get's overlapped by source-view content.
* Moving the mouse inside the ToC pane causes unexpected display of link info details (Go to "...") in the underlying pageview when there's a link on the page at the same relative y-pos of the mouse pointer.

introt commented:
No problem! Really excited for this feature, as it speeds up my current workflow quite a bit.

I'm in the middle of my Monday here, but was able to get some insight as well as a logging error for good measure :D

Here's my quick'n'dirty logging setup, stripped down to essentials:

```python3
    def _update_link_tag(self, elt, source, target, oldhref):
        if oldhref.rel == HREF_REL_ABSOLUTE: # prefer to keep absolute links
            newhref = HRef(HREF_REL_ABSOLUTE, target.name)
        elif source == target and oldhref.anchor:
            newhref = HRef(HREF_REL_FLOATING, '', oldhref.anchor)
        else:
            newhref = self.pages.create_link(source, target)
        logger.debug("newhref:", newhref.parts(), newhref.to_wiki_link())

        newhref.anchor = oldhref.anchor

        link = newhref.to_wiki_link()

        logger.debug("elt:", elt.gettext(), "oldref:", oldhref.to_wiki_link(), oldhref.parts())
        if elt.gettext() == elt.get('href'):
            elt[:] = [link]
            logger.debug("gettext==href")
        elif elt.gettext() == oldhref.parts()[-1] and len(elt) == 1:
            # We are using short links and the link text was short link
            # and there were no sub-node (like bold text) that would be cancelled.
            # Related to 'short_links' but not checking the property here.
            logger.debug("gettext==last part")
            short = newhref.parts()[-1]
            if newhref.anchor:
                logger.debug("newhref.anchor")
                short += '#' + newhref.anchor
            elt[:] = [short]  # 'Journal:2020:01:20' -> '20'
        else:
            logger.debug("neither")

        elt.set('href', link)

        return elt
```

Relevant log snippets with PR:
```python3
    logger.debug("newhref:", newhref.parts(), newhref.to_wiki_link())
Message: 'newhref:'
Arguments: ([], '#subsub')
--- Logging error ---
```

```python3
    logger.debug("elt:", elt.gettext(), "oldref:", oldhref.to_wiki_link(), oldhref.parts())
Message: 'elt:'
Arguments: ('subsub', 'oldref:', 'test:subtest:subsub#subsub', ['test', 'subtest', 'subsub'])
DEBUG: gettext==last part
```

The same log parts on `develop`:

```python3
    logger.debug("newhref:", newhref.parts(), newhref.to_wiki_link())
Message: 'newhref:'
Arguments: ([], '#subsub')
--- Logging error ---
```

```python3
    logger.debug("elt:", elt.gettext(), "oldref:", oldhref.to_wiki_link(), oldhref.parts())
Message: 'elt:'
Arguments: ('subsub\n', 'oldref:', 'test:subtest:subsub#subsub', ['test', 'subtest', 'subsub'])
DEBUG: neither
```

See the difference? I may not have figured it out if it weren't for the logging error, as I started by logging `elt`, not `elt.gettext()` and the line wrapped very similarly. I haven't the slightest clue about where `elt` comes from, but looking through your changes, one of them seems especially _sus_, as the kids would say.

introt commented on zim/gui/pageview/__init__.py:
```diff
@@ -1390,18 +1390,18 @@ def _get_close_anchor_or_object_id(self, iter):
 		line_end = line_start.copy()
 		line_end.forward_line()
-		return line_start.get_text(line_end)
+		return line_start.get_text(line_end).strip()
```
This is the breaking change, as confirmed by the issue getting resolved by taking it out. I agree that stripping should totally be applied here, but as we've seen, it does expose at least one bug elsewhere.

So: do we leave it out, or brace ourselves for a bunch of issues that may creep up sooner or later? I reckon it's up to Jaap to decide. Either way, changes need to be made in one direction or the other.

roland-ruedenauer replied:
Hi,
please check if  the change I commited solves the issue you ran into.

introt commented on zim/notebook/notebook.py:
```diff
@@ -835,7 +835,7 @@ def _update_link_tag(self, elt, source, target, oldhref):
 		if elt.gettext() == elt.get('href'):
 			elt[:] = [link]
-		elif elt.gettext() == oldhref.parts()[-1] and len(elt) == 1:
+		elif oldhref.parts() and elt.gettext() == oldhref.parts()[-1] and len(elt) == 1:
```
I think you meant to type `newhref` instead of `oldhref` here - that fixes the issue.

roland-ruedenauer replied:
I added a check for `newhref` being empty too (the case you ran into). While testing yesterday I had an issue related to `oldhref` which was obviously a different problem.
This would also mean I was not able to reproduce your initial problem. Anyway, the parts list maybe empty and we have to check for this case.

introt commented:
Forgot to ask about this yesterday - did you look into implementing drag'n'drop? Would speed up the flow further by not having to to open the menu and triggering paste separately. Don't think we're in a hurry here - new features w/ possible side effects may need to wait till 0.74 is "stable" before getting merged into develop for inclusion in 0.75, if I've read Jaap correctly.

roland-ruedenauer commented:
I did not have a look into also supporting drag'n drop for now. I can have a look into this too but it could take a while until I find some time.

introt commented on zim/notebook/notebook.py:
```diff
@@ -835,11 +835,11 @@ def _update_link_tag(self, elt, source, target, oldhref):
 			# and there were no sub-node (like bold text) that would be cancelled.
 			# Related to 'short_links' but not checking the property here.
-			short = newhref.parts()[-1]
+			short = newhref.parts()[-1] if newhref.parts() else ''
```
This makes pasting the heading produce a link like `#this` instead of `this` like all the other headings do. Imo they should all behave the same way; if that way is `this`, it can be achieved with `elif oldhref.parts() and newhref.parts() and ...` on line 838

introt commented:
> I did not have a look into also supporting drag'n drop for now. I can have a look into this too but it could take a while until I find some time.

Jaap seems to be making destructive changes in develop again (eg. migrating to `newfs`) so it might now make sense to get this wrapped up && merged before looking into d'n'd. Up to you, I really appreciate your efforts either way!

roland-ruedenauer commented:
Hi @introt, I rebased the PR after the many changes applied lately and fixed the failing tests.
Can you check if you copy&paste problem is still present?

Note also the fix Jaap commit in 88eef741356c9e0e47eb698d58302ac3dc853c79.

I also had a look into implementing drag-and-drop in the meantime. It's working for me principally - meaning I get the expected link created at the drop location. What's not ok, is the visual feedback indicating the drop position and the mouse pointer changing when entering/leaving the text area. It works for a short time but then all a sudden get's not updated correctly anymore.
This might be caused by zim itself as I see it working in a tiny test application. I don't see these issues resolved soon and excluded my DnD changes from the PR. It's better to tackle this separately.

introt reviewed (approved):
Thanks for the great work @roland-ruedenauer! Copying headers now works as expected both between pages and on the same page. I can also confirm that your changes fix  #1730 and #1769, including Move Selected Text. Code looks good to me based on a cursory scroll, I'll have to leave the API changes for Jaap to judge.

Re: drag'n'drop, could you leave the code up somewhere ~for me to use~ to aid with hunting down the issue? Thanks!

jaap-karssenberg commented on zim/gui/server.py:
```diff
@@ -132,7 +132,7 @@ def start(self):
 			uri = self.notebookcombobox.get_notebook()
 			if uri:
-				notebook, x = build_notebook(NotebookInfo(uri))
+				notebook, _ = build_notebook(NotebookInfo(uri))
```
idem dito - please avoid these style changes

jaap-karssenberg commented on zim/main/__init__.py:
```diff
@@ -39,12 +38,14 @@
 
 	usagehelp = '''\
-usage: zim [OPTIONS] [NOTEBOOK [PAGE]]
+usage: zim [OPTIONS] [NOTEBOOK [PAGE [ANCHOR]]]
```
I would strongly prefer the ANCHOR argument to be provided via an option. I would suggest "--id" as the user interface also refers to these as "id" rather than "anchor".  

I'm OK with allowing the anchor as part of the PAGE argument, since the "#" is reserved and not allowed in page names.

jaap-karssenberg commented on zim/main/__init__.py:
```diff
@@ -137,75 +138,82 @@ def get_default_or_only_notebook(self):
 	'''Class implementing the C{--gui} command and run the gtk interface'''
 
-	arguments = ('[NOTEBOOK]', '[PAGE]')
+	arguments = ('[NOTEBOOK]', '[PAGE]', '[ANCHOR]')
```
Idem dito, prefer "--id" for the anchor argument

jaap-karssenberg commented on zim/main/__init__.py:
```diff
@@ -658,6 +669,7 @@ def __init__(self):
 		self._log_started = False
 		self._standalone = False
 		self._windows = set()
+		self._arguments = None
```
What is the purpose of this attribute ? Can't find back where it is used.

jaap-karssenberg commented on zim/notebook/page.py:
```diff
@@ -372,6 +372,10 @@ def __str__(self):
 	def parts(self):
 		return self.names.split(':') if self.names else []
 
+	def short_name(self):
```
Good addition! Cleans up a bit of jargon in different places.

jaap-karssenberg commented on zim/plugins/tableofcontents.py:
```diff
@@ -76,7 +75,10 @@ def get_headings(parsetree, include_hr):
 			level = int(t[1]['level'])
 			text = tokens_to_text(
-						collect_untill_end_token(tokens, HEADING) ).strip()
+				filter_token(
```
This fix is already solved another way - please see commit 5b87aa73e3c96772c81dab692a33ef1acbc2c42e

jaap-karssenberg commented on zim/tokenparser.py:
```diff
@@ -21,9 +21,16 @@ class EndOfTokenListError(AssertionError):
 
 
-def collect_untill_end_token(token_iter, tag):
+def collect_until_end_token(token_iter, tag):
```
Good catch on the typo in the function name.

Would prefer to revers this function to the implementation without a generator. Function is small and low level so called often in parsing so rather keep it a single function call. In this better to have a bit of redundancy in functions

jaap-karssenberg commented on zim/tokenparser.py:
```diff
@@ -32,12 +39,26 @@ def collect_untill_end_token(token_iter, tag):
 
-	return tokens
+
+def filter_token(token_iter, token):
```
Good addition to the toolkit. Please leave this function in, even if the patch doesn't need it anymore (due to fix in 5b87aa73e3c96772c81dab692a33ef1acbc2c42e).

Thanks!

jaap-karssenberg commented on zim/plugins/tableofcontents.py:
```diff
@@ -396,8 +424,30 @@ def on_populate_popup(self, treeview, menu):
+		sep = Gtk.SeparatorMenuItem()
+		menu.append(sep)
+
+		item = Gtk.MenuItem.new_with_mnemonic(_("Copy _link to this location"))
```
Minor: for translation strings, please add a comment like "# T: purpose of string" this is picked up as info for translators

jaap-karssenberg commented on zim/plugins/tableofcontents.py:
```diff
@@ -396,8 +424,30 @@ def on_populate_popup(self, treeview, menu):
+			item.set_sensitive(False)
+
+		# open in new window
+		item = Gtk.MenuItem.new_with_mnemonic(_('Open in New _Window'))
```
Idem dito: please add translation comment

jaap-karssenberg reviewed (changes requested):
Hi Roland!

Sorry it took me some time to get around this review. Due to the whole covid thing lot of spare time got sucked away into child care and work ..

Added a bunch of comments, but overall I very much like the improvements you are making 👍 

The main blocking item for me is the commandline syntax, but open for discussion. You might want to file separate merge request for the ToC and the commandline arguments. That allows to discuss separate. But also OK to keep them together as long as it are two commits.

Thanks !

Jaap

introt commented:
I've split token parsing and HRef.short_name into new PRs #1947 and #1948; these include the requested changes (as well as extra fixes & related tests in the latter).

Besides those, this PR includes the following which should be split off before closing:

- [ ] linking related UX in ToC pane (in case @roland-ruedenauer stays MIA)
  - [ ] the very welcome addition of `strip()` to `get_heading_text`
- [ ] the `anchor` arg, which imo should just be included as part of the `page` (which itself should be renamed to `url` or `link` or something)
  * in addition to 458688c2042602ba4a0c94dab8b1e11f9008911d, this would also benefit from having something "proper check link syntax, including achor part" --> Discuss at #1942

introt commented:
Keeping this open for the anchor stuff, ToC functionality is now in #1952

introt commented:
Aforementioned anchor stuff handled by f2f6565b90f01f30978771d8d2f1b85628d475ce
